### DISCUSSION TITLE
Return link after completion of experiment

### DISCUSSION THREAD

**Discussion by TomRui** [jspsych/jsPsych#1460] [Q&A] [answered]
Hi all!

Is there a way to automatically redirect participants to an URL after they complete a task in jsPsych, so that they can be redirected to e.g. Qualtrics again?

I found [this](https://discourse.psychopy.org/t/how-to-provide-return-link-in-jspsych-for-participants-recruited-through-prolific/10104) older topic on the subject but no definite answer was given, only a workaround which the author himself did not like either.

jsPsych looks very promising but this is a key feature that we'd rely on in our lab. Thanks a lot for any guidance you can provide!

Edit: I'd plan to host it on either Pavlovia or Cognition.run if that's relevant

**Comment by jodeleeuw** [comment, marked as answer]
Hi @TomRui,

Yep, you can use a couple different methods. 

My preferred approach is to display the link as clickable text so participants know that they are being redirected. You can use an `html-keyboard-response` trial to do this:

```js
var redirect_trial = {
	type: 'html-keyboard-response',
	stimulus: '<p>You have completed the experiment.</p><p><a href="url-to-redirect-to">Please click here to return to Qualtrics.</a></p>',
	choices: jsPsych.NO_KEYS
}
```

You could also redirect the participant manually via a call to `window.location`. For example:

```js
jsPsych.init({
	timeline: [...],
	on_finish: function() { window.location = "url-to-redirect-to"; }
})
```

**Reply by TomRui** [reply]
Hi @jodeleeuw, 

Thanks a lot for the quick reply! In the [topic](https://discourse.psychopy.org/t/how-to-provide-return-link-in-jspsych-for-participants-recruited-through-prolific/10104) about this their fear is that the data won't be saved this way.

Is this unfounded, or does this differ per hosting platform? If so, I'll redirect my question to the peeps at Pavlovia or Cognition.run :)

Best,
Tom

**Reply by jodeleeuw** [reply]
This will differ by hosting platform. The issue is waiting for any async requests to the server to complete before redirecting. Ideally the platform would provide a callback function that is executed once the data are saved. Cognition.run avoids this problem by saving the data automatically on every trial. I'm not sure if an alternative solution exists for Pavlovia.

I use a self-hosted web server to run my experiments, but the approach that I use is to send the data to the server prior to the debrief screen using a `call-function` plugin. At the end of the debrief I have a link to return folks to the site that they came from. I haven't had issues with data loss using this strategy, which I think means that on my server I only need to stall people for a few seconds before the request is safely completed.

**Reply by TomRui** [reply]
Clear, thanks for the info!
